// @sanity/desk-tool/lib/panes/document/DocumentPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.DocumentPane = void 0;
var _hooks = require("@sanity/base/hooks");
var _components = require("@sanity/base/components");
var _changeIndicators = require("@sanity/base/change-indicators");
var _initialValueTemplates = require("@sanity/base/initial-value-templates");
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _paths = require("@sanity/util/paths");
var _internal = require("@sanity/form-builder/_internal");
var _internal2 = require("@sanity/base/_internal");
var _deskTool = require("../../contexts/deskTool");
var _paneRouter = require("../../contexts/paneRouter");
var _pane = require("../../components/pane");
var _usePaneLayout2 = require("../../components/pane/usePaneLayout");
var _error = require("../error");
var _loading = require("../loading");
var _changesPanel = require("./changesPanel");
var _documentPanel = require("./documentPanel");
var _DocumentOperationResults = require("./DocumentOperationResults");
var _keyboardShortcuts = require("./keyboardShortcuts");
var _statusBar = require("./statusBar");
var _useDocumentPane2 = require("./useDocumentPane");
var _DocumentPaneProvider = require("./DocumentPaneProvider");
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = null == arr ? null : "undefined" != typeof Symbol && arr[Symbol.iterator] || arr["@@iterator"]; if (null != _i) { var _s, _e, _x, _r, _arr = [], _n = !0, _d = !1; try { if (_x = (_i = _i.call(arr)).next, 0 === i) { if (Object(_i) !== _i) return; _n = !1; } else for (; !(_n = (_s = _x.call(_i)).done) && (_arr.push(_s.value), _arr.length !== i); _n = !0); } catch (err) { _d = !0, _e = err; } finally { try { if (!_n && null != _i.return && (_r = _i.return(), Object(_r) !== _r)) return; } finally { if (_d) throw _e; } } return _arr; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var DOCUMENT_PANEL_MIN_WIDTH = 320;
var DOCUMENT_PANEL_INITIAL_MIN_WIDTH = 600;
var CHANGES_PANEL_MIN_WIDTH = 320;
var DIALOG_PROVIDER_POSITION = [
// We use the `position: fixed` for dialogs on narrow screens (< 512px).
'fixed',
// And we use the `position: absolute` strategy (within panes) on wide screens.
'absolute'];
var StyledChangeConnectorRoot = (0, _styledComponents.default)(_changeIndicators.ChangeConnectorRoot)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  flex: 1;\n  display: flex;\n  min-height: 0;\n  min-width: 0;\n"])));
var DocumentPane = /*#__PURE__*/(0, _react.memo)(function DocumentPane(props) {
  var _routerPanesState;
  var paneRouter = (0, _paneRouter.usePaneRouter)();
  var options = usePaneOptions(props.pane.options, paneRouter.params);
  var _useDocumentType = (0, _hooks.useDocumentType)(options.id, options.type),
    documentType = _useDocumentType.documentType,
    isDocumentLoaded = _useDocumentType.isLoaded;
  var _useUnstableTemplateP = (0, _hooks.unstable_useTemplatePermissions)((0, _internal2.getNewDocumentOptions)()),
    _useUnstableTemplateP2 = _slicedToArray(_useUnstableTemplateP, 2),
    templatePermissions = _useUnstableTemplateP2[0],
    isTemplatePermissionsLoading = _useUnstableTemplateP2[1];
  var isLoaded = isDocumentLoaded && !isTemplatePermissionsLoading;
  var providerProps = (0, _react.useMemo)(() => {
    return isLoaded && documentType && options.type !== documentType ? mergeDocumentType(props, options, documentType) : props;
  }, [props, documentType, isLoaded, options]);
  var ReferenceChildLink = paneRouter.ReferenceChildLink,
    handleEditReference = paneRouter.handleEditReference,
    groupIndex = paneRouter.groupIndex,
    routerPanesState = paneRouter.routerPanesState;
  var childParams = ((_routerPanesState = routerPanesState[groupIndex + 1]) === null || _routerPanesState === void 0 ? void 0 : _routerPanesState[0].params) || {};
  var routerPanesStateLength = routerPanesState.length;
  var parentRefPath = childParams.parentRefPath;
  var activePath = (0, _react.useMemo)(() => {
    return parentRefPath ? {
      path: (0, _paths.fromString)(parentRefPath),
      state:
      // eslint-disable-next-line no-nested-ternary
      groupIndex >= routerPanesStateLength - 1 ? 'none' : groupIndex >= routerPanesStateLength - 2 ? 'selected' : 'pressed'
    } : {
      path: [],
      state: 'none'
    };
  }, [parentRefPath, groupIndex, routerPanesStateLength]);
  if (options.type === '*' && !isLoaded) {
    return /*#__PURE__*/_react.default.createElement(_loading.LoadingPane, {
      flex: 2.5,
      minWidth: 320,
      paneKey: props.paneKey,
      title: "Loading document\u2026"
    });
  }
  if (!documentType) {
    return /*#__PURE__*/_react.default.createElement(_error.ErrorPane, {
      flex: 2.5,
      minWidth: 320,
      paneKey: props.paneKey,
      title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "The document was not found")
    }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
      space: 4
    }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
      as: "p"
    }, "The document type is not defined, and a document with the ", /*#__PURE__*/_react.default.createElement("code", null, options.id), ' ', "identifier could not be found.")));
  }
  return /*#__PURE__*/_react.default.createElement(_DocumentPaneProvider.DocumentPaneProvider, providerProps, /*#__PURE__*/_react.default.createElement(_internal.Unstable_ReferenceInputOptionsProvider, {
    EditReferenceLinkComponent: ReferenceChildLink,
    onEditReference: handleEditReference,
    initialValueTemplateItems: templatePermissions,
    activePath: activePath
  }, /*#__PURE__*/_react.default.createElement(InnerDocumentPane, null)));
});
exports.DocumentPane = DocumentPane;
function usePaneOptions(options) {
  var params = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
  return (0, _react.useMemo)(() => {
    // The document type is provided, so return
    if (options.type && options.type !== '*') {
      return options;
    }

    // Attempt to derive document type from the template configuration
    var templateName = options.template || params.template;
    var template = templateName ? (0, _initialValueTemplates.getTemplateById)(templateName) : undefined;
    var documentType = template === null || template === void 0 ? void 0 : template.schemaType;

    // No document type was found in a template
    if (!documentType) {
      return options;
    }

    // The template provided the document type, so modify the pane’s `options` property
    return _objectSpread(_objectSpread({}, options), {}, {
      type: documentType
    });
  }, [options, params.template]);
}
function mergeDocumentType(props, options, documentType) {
  return _objectSpread(_objectSpread({}, props), {}, {
    pane: _objectSpread(_objectSpread({}, props.pane), {}, {
      options: _objectSpread(_objectSpread({}, options), {}, {
        type: documentType
      })
    })
  });
}
function InnerDocumentPane() {
  var _useDocumentPane = (0, _useDocumentPane2.useDocumentPane)(),
    changesOpen = _useDocumentPane.changesOpen,
    documentSchema = _useDocumentPane.documentSchema,
    documentType = _useDocumentPane.documentType,
    handleFocus = _useDocumentPane.handleFocus,
    handleHistoryOpen = _useDocumentPane.handleHistoryOpen,
    handleKeyUp = _useDocumentPane.handleKeyUp,
    inspectOpen = _useDocumentPane.inspectOpen,
    paneKey = _useDocumentPane.paneKey,
    value = _useDocumentPane.value,
    totalReferenceCount = _useDocumentPane.totalReferenceCount;
  var _useDeskTool = (0, _deskTool.useDeskTool)(),
    features = _useDeskTool.features;
  var _usePaneLayout = (0, _usePaneLayout2.usePaneLayout)(),
    layoutCollapsed = _usePaneLayout.collapsed;
  var zOffsets = (0, _components.useZIndex)();
  var _useState = (0, _react.useState)(null),
    _useState2 = _slicedToArray(_useState, 2),
    rootElement = _useState2[0],
    setRootElement = _useState2[1];
  var _useState3 = (0, _react.useState)(null),
    _useState4 = _slicedToArray(_useState3, 2),
    footerElement = _useState4[0],
    setFooterElement = _useState4[1];
  var _useState5 = (0, _react.useState)(null),
    _useState6 = _slicedToArray(_useState5, 2),
    actionsBoxElement = _useState6[0],
    setActionsBoxElement = _useState6[1];
  var footerRect = (0, _ui.useElementRect)(footerElement);
  var footerH = footerRect === null || footerRect === void 0 ? void 0 : footerRect.height;
  var documentIsReferenced = totalReferenceCount !== undefined && totalReferenceCount > 0;
  var documentPanel = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_documentPanel.DocumentPanel, {
    footerHeight: footerH || null,
    rootElement: rootElement,
    isInspectOpen: inspectOpen
  }), [footerH, rootElement, inspectOpen]);
  var footer = (0, _react.useMemo)(() => /*#__PURE__*/_react.default.createElement(_pane.PaneFooter, {
    isReferencedDocument: documentIsReferenced,
    ref: setFooterElement
  }, /*#__PURE__*/_react.default.createElement(_statusBar.DocumentStatusBar, {
    actionsBoxRef: setActionsBoxElement
  })), [documentIsReferenced]);
  var changesPanel = (0, _react.useMemo)(() => {
    if (!features.reviewChanges) return null;
    if (!changesOpen) return null;
    return /*#__PURE__*/_react.default.createElement(_ui.BoundaryElementProvider, {
      element: rootElement
    }, /*#__PURE__*/_react.default.createElement(_changesPanel.ChangesPanel, null));
  }, [changesOpen, features.reviewChanges, rootElement]);
  var children = (0, _react.useMemo)(() => {
    if (!documentSchema) {
      return /*#__PURE__*/_react.default.createElement(_error.ErrorPane, {
        flex: 2.5,
        minWidth: 320,
        paneKey: paneKey,
        title: /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, "Unknown document type: ", /*#__PURE__*/_react.default.createElement("code", null, documentType)),
        tone: "caution"
      }, /*#__PURE__*/_react.default.createElement(_ui.Stack, {
        space: 4
      }, documentType && /*#__PURE__*/_react.default.createElement(_ui.Text, {
        as: "p"
      }, "This document has the schema type ", /*#__PURE__*/_react.default.createElement("code", null, documentType), ", which is not defined as a type in the local content studio schema."), !documentType && /*#__PURE__*/_react.default.createElement(_ui.Text, {
        as: "p"
      }, "This document does not exist, and no schema type was specified for it."), __DEV__ && value && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.Text, {
        as: "p"
      }, "Here is the JSON representation of the document:"), /*#__PURE__*/_react.default.createElement(_ui.Card, {
        padding: 3,
        overflow: "auto",
        radius: 2,
        shadow: 1,
        tone: "inherit"
      }, /*#__PURE__*/_react.default.createElement(_ui.Code, {
        language: "json",
        size: [1, 1, 2]
      }, JSON.stringify(value, null, 2))))));
    }
    return /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement(_ui.DialogProvider, {
      position: DIALOG_PROVIDER_POSITION,
      zOffset: zOffsets.portal
    }, /*#__PURE__*/_react.default.createElement(_ui.Flex, {
      direction: "column",
      flex: 1,
      height: layoutCollapsed ? undefined : 'fill'
    }, /*#__PURE__*/_react.default.createElement(StyledChangeConnectorRoot, {
      "data-testid": "change-connector-root",
      isReviewChangesOpen: changesOpen,
      onOpenReviewChanges: handleHistoryOpen,
      onSetFocus: handleFocus
    }, documentPanel, changesPanel))), footer, /*#__PURE__*/_react.default.createElement(_DocumentOperationResults.DocumentOperationResults, null));
  }, [changesOpen, changesPanel, documentPanel, documentSchema, documentType, footer, handleFocus, handleHistoryOpen, layoutCollapsed, paneKey, value, zOffsets.portal]);
  var currentMinWidth = changesOpen ? DOCUMENT_PANEL_INITIAL_MIN_WIDTH + CHANGES_PANEL_MIN_WIDTH : DOCUMENT_PANEL_INITIAL_MIN_WIDTH;
  var minWidth = changesOpen ? DOCUMENT_PANEL_MIN_WIDTH + CHANGES_PANEL_MIN_WIDTH : DOCUMENT_PANEL_MIN_WIDTH;
  return /*#__PURE__*/_react.default.createElement(_keyboardShortcuts.DocumentActionShortcuts, {
    actionsBoxElement: actionsBoxElement,
    currentMinWidth: currentMinWidth,
    "data-testid": "document-pane",
    flex: 2.5,
    id: paneKey,
    minWidth: minWidth,
    onKeyUp: handleKeyUp,
    rootRef: setRootElement
  }, children);
}
