// @sanity/structure/lib/MenuItemGroup.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MenuItemGroupBuilder = void 0;
exports.maybeSerializeMenuItemGroup = maybeSerializeMenuItemGroup;
var _SerializeError = require("./SerializeError");
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
function maybeSerializeMenuItemGroup(item, index, path) {
  return item instanceof MenuItemGroupBuilder ? item.serialize({
    path,
    index
  }) : item;
}
class MenuItemGroupBuilder {
  constructor(spec) {
    _defineProperty(this, "_id", void 0);
    _defineProperty(this, "_title", void 0);
    this._id = spec ? spec.id : '';
    this._title = spec ? spec.title : '';
  }
  id(id) {
    return new MenuItemGroupBuilder({
      id,
      title: this._title
    });
  }
  getId() {
    return this._id;
  }
  title(title) {
    return new MenuItemGroupBuilder({
      id: this._id,
      title
    });
  }
  getTitle() {
    return this._title;
  }
  serialize() {
    var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
      path: []
    };
    var _id = this._id,
      _title = this._title;
    if (!_id) {
      throw new _SerializeError.SerializeError('`id` is required for a menu item group', options.path, options.index, _title).withHelpUrl(_SerializeError.HELP_URL.ID_REQUIRED);
    }
    if (!_title) {
      throw new _SerializeError.SerializeError('`title` is required for a menu item group', options.path, _id).withHelpUrl(_SerializeError.HELP_URL.TITLE_REQUIRED);
    }
    return {
      id: _id,
      title: _title
    };
  }
}
exports.MenuItemGroupBuilder = MenuItemGroupBuilder;
