// @sanity/desk-tool/lib/panes/loading/LoadingPane.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.LoadingPane = void 0;
var _ui = require("@sanity/ui");
var _react = _interopRequireWildcard(require("react"));
var _styledComponents = _interopRequireDefault(require("styled-components"));
var _Delay = require("../../components/Delay");
var _pane = require("../../components/pane");
var _templateObject;
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }
function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }
function _slicedToArray(arr, i) { return _arrayWithHoles(arr) || _iterableToArrayLimit(arr, i) || _unsupportedIterableToArray(arr, i) || _nonIterableRest(); }
function _nonIterableRest() { throw new TypeError("Invalid attempt to destructure non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }
function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }
function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) arr2[i] = arr[i]; return arr2; }
function _iterableToArrayLimit(arr, i) { var _i = null == arr ? null : "undefined" != typeof Symbol && arr[Symbol.iterator] || arr["@@iterator"]; if (null != _i) { var _s, _e, _x, _r, _arr = [], _n = !0, _d = !1; try { if (_x = (_i = _i.call(arr)).next, 0 === i) { if (Object(_i) !== _i) return; _n = !1; } else for (; !(_n = (_s = _x.call(_i)).done) && (_arr.push(_s.value), _arr.length !== i); _n = !0); } catch (err) { _d = !0, _e = err; } finally { try { if (!_n && null != _i.return && (_r = _i.return(), Object(_r) !== _r)) return; } finally { if (_d) throw _e; } } return _arr; } }
function _arrayWithHoles(arr) { if (Array.isArray(arr)) return arr; }
function _taggedTemplateLiteral(strings, raw) { if (!raw) { raw = strings.slice(0); } return Object.freeze(Object.defineProperties(strings, { raw: { value: Object.freeze(raw) } })); }
var DELAY = false;
var DEFAULT_MESSAGE = 'Loading…';
var Content = (0, _styledComponents.default)(_ui.Flex)(_templateObject || (_templateObject = _taggedTemplateLiteral(["\n  opacity: 0;\n  transition: opacity 200ms;\n\n  &[data-mounted] {\n    opacity: 1;\n  }\n"])));

/**
 * @internal
 */
var LoadingPane = /*#__PURE__*/(0, _react.memo)(props => {
  var _props$delay = props.delay,
    delay = _props$delay === void 0 ? 300 : _props$delay,
    flex = props.flex,
    messageProp = props.message,
    minWidth = props.minWidth,
    paneKey = props.paneKey,
    path = props.path,
    selected = props.selected,
    title = props.title,
    tone = props.tone;
  var resolvedMessage = (0, _react.useMemo)(() => {
    if (typeof messageProp === 'function') {
      return messageProp(path ? path.split(';') : []);
    }
    return messageProp;
  }, [messageProp, path]);
  var _useState = (0, _react.useState)(() => {
      if (typeof resolvedMessage === 'string') return resolvedMessage;
      return DEFAULT_MESSAGE;
    }),
    _useState2 = _slicedToArray(_useState, 2),
    currentMessage = _useState2[0],
    setCurrentMessage = _useState2[1];
  (0, _react.useEffect)(() => {
    if (typeof resolvedMessage !== 'object') return undefined;
    if (typeof resolvedMessage.subscribe === 'function') return undefined;
    var sub = resolvedMessage.subscribe(setCurrentMessage);
    return () => sub.unsubscribe();
  }, [resolvedMessage]);
  var _useState3 = (0, _react.useState)(null),
    _useState4 = _slicedToArray(_useState3, 2),
    contentElement = _useState4[0],
    setContentElement = _useState4[1];
  var _useState5 = (0, _react.useState)(false),
    _useState6 = _slicedToArray(_useState5, 2),
    mounted = _useState6[0],
    setMounted = _useState6[1];
  (0, _react.useEffect)(() => {
    if (!contentElement) return undefined;
    return (0, _ui._raf2)(() => setMounted(true));
  }, [contentElement]);
  var content = /*#__PURE__*/_react.default.createElement(Content, {
    align: "center",
    "data-mounted": mounted ? '' : undefined,
    direction: "column",
    height: "fill",
    justify: "center",
    ref: setContentElement
  }, /*#__PURE__*/_react.default.createElement(_ui.Spinner, {
    muted: true
  }), (title || currentMessage) && /*#__PURE__*/_react.default.createElement(_ui.Box, {
    marginTop: 3
  }, /*#__PURE__*/_react.default.createElement(_ui.Text, {
    align: "center",
    muted: true,
    size: 1
  }, title || currentMessage)));
  return /*#__PURE__*/_react.default.createElement(_pane.Pane, {
    flex: flex,
    id: paneKey,
    minWidth: minWidth,
    selected: selected,
    tone: tone
  }, /*#__PURE__*/_react.default.createElement(_pane.PaneContent, null, DELAY ? /*#__PURE__*/_react.default.createElement(_Delay.Delay, {
    ms: delay
  }, content) : content));
});
exports.LoadingPane = LoadingPane;
LoadingPane.displayName = 'LoadingPane';
