// @sanity/base/lib/__legacy/@sanity/components/tabs/TabList.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _TabList = _interopRequireDefault(require("./TabList.css"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
// @todo: refactor to functional component
class TabList extends _react.default.PureComponent {
  constructor() {
    super(...arguments);
    _defineProperty(this, "state", {
      focusedTabIdx: -1
    });
    _defineProperty(this, "handleTabFocus", tabIdx => {
      this.setState({
        focusedTabIdx: tabIdx
      });
    });
    _defineProperty(this, "handleKeyDown", evt => {
      var numTabs = this.props.children.length;
      if (evt.key === 'ArrowLeft') {
        this.setState(state => {
          var focusedTabIdx = state.focusedTabIdx < 1 ? numTabs - 1 : state.focusedTabIdx - 1;
          return {
            focusedTabIdx
          };
        });
      }
      if (evt.key === 'ArrowRight') {
        this.setState(state => {
          var focusedTabIdx = (state.focusedTabIdx + 1) % numTabs;
          return {
            focusedTabIdx
          };
        });
      }
    });
  }
  render() {
    var focusedTabIdx = this.state.focusedTabIdx;
    var children = this.props.children.map((child, idx) => {
      return /*#__PURE__*/_react.default.cloneElement(child, {
        isFocused: focusedTabIdx === idx,
        onFocus: () => this.handleTabFocus(idx)
      });
    });
    return /*#__PURE__*/_react.default.createElement("div", {
      className: _TabList.default.root,
      onKeyDown: this.handleKeyDown,
      role: "tablist"
    }, /*#__PURE__*/_react.default.createElement("div", null, children));
  }
}
exports.default = TabList;
